Code/Framework/AzToolsFramework/AzToolsFramework/ViewportSelection/ViewportEditorModeTracker.cpp

```diff
@@ -0,0 +1,121 @@
+        }
+    }
+
+    bool ViewportEditorModes::IsModeActive(ViewportEditorMode mode) const
+    {
+        return m_editorModes[static_cast<AZ::u32>(mode)];
+    }
+
+    void ViewportEditorModeTracker::RegisterInterface()
+    {
+        if (AZ::Interface<ViewportEditorModeTrackerInterface>::Get() == nullptr)
+        {
+            AZ::Interface<ViewportEditorModeTrackerInterface>::Register(this);
+        }
+    }
+
+    void ViewportEditorModeTracker::UnregisterInterface()
+    {
+        if (AZ::Interface<ViewportEditorModeTrackerInterface>::Get() != nullptr)
+        {
+            AZ::Interface<ViewportEditorModeTrackerInterface>::Unregister(this);
+        }
+    }
+
+    void ViewportEditorModeTracker::EnterMode(const ViewportEditorModeInfo& viewportEditorModeInfo, ViewportEditorMode mode)
+    {
+        auto& editorModeStates = m_viewportEditorModeStates[viewportEditorModeInfo.m_id];
+        AZ_Warning(
+                ViewportEditorModeLogWindow, !editorModeStates.IsModeActive(mode),
+                AZStd::string::format(
+                    "Duplicate call to EnterMode for mode '%u' on id '%i'", static_cast<AZ::u32>(mode), viewportEditorModeInfo.m_id).c_str());
+        editorModeStates.SetModeActive(mode);
+        ViewportEditorModeNotificationsBus::Event(
+            viewportEditorModeInfo.m_id, &ViewportEditorModeNotificationsBus::Events::OnEditorModeEnter, editorModeStates, mode);
+    }
+
+    void ViewportEditorModeTracker::ExitMode(const ViewportEditorModeInfo& viewportEditorModeInfo, ViewportEditorMode mode)
+    {
+        ViewportEditorModes* editorModeStates = nullptr;
+        if (m_viewportEditorModeStates.count(viewportEditorModeInfo.m_id))
+        {
+            editorModeStates = &m_viewportEditorModeStates.at(viewportEditorModeInfo.m_id);
+            AZ_Warning(
+                ViewportEditorModeLogWindow, editorModeStates->IsModeActive(mode),
+                AZStd::string::format(
+                    "Duplicate call to ExitMode for mode '%u' on id '%i'", static_cast<AZ::u32>(mode), viewportEditorModeInfo.m_id).c_str());
+        }
+        else
+        {
+            AZ_Warning(
+                ViewportEditorModeLogWindow, false, "Call to ExitMode for mode '%u' on id '%i' without precursor call to EnterMode",
+                static_cast<AZ::u32>(mode), viewportEditorModeInfo.m_id);
+
+            editorModeStates = &m_viewportEditorModeStates[viewportEditorModeInfo.m_id];
+        }
+
+        editorModeStates->SetModeInactive(mode);
+        ViewportEditorModeNotificationsBus::Event(
+            viewportEditorModeInfo.m_id, &ViewportEditorModeNotificationsBus::Events::OnEditorModeExit, *editorModeStates, mode);
+    }
+
+    const ViewportEditorModesInterface* ViewportEditorModeTracker::GetViewportEditorModes(const ViewportEditorModeInfo& viewportEditorModeInfo) const
+    {
+        if (auto editorModeStates = m_viewportEditorModeStates.find(viewportEditorModeInfo.m_id);
```

Comment:
nit: Should this variable (and others in the file) just be `editorModes` instead?

Reply:
fixed 👍